### DISCUSSION TITLE
Global DDL lock while export

### DISCUSSION THREAD

**Discussion by viswaprsnl** [mydumper/mydumper#1284] [Q&A] [answered]
I am using mydumper 0.15 . I see "Acquiring DDL lock" while running exports .Does that mean my entire DB is locked while my export is running. is this statement true ?

**Comment by davidducos** [comment, marked as answer]
We sent DDL locks when is possible to get a consistent backup. DDL lock doesn't block your entire DB, just DDL statements like CREATE TABLE, DROP TABLE, ALTER TABLE, etc. 

The scenario that we tried to avoid is when you DROP a table that mydumper considers that needs to backup.

**Comment by viswaprsnl** [comment]
@davidducos  Thanks , so if i run a parallel import of a different database (DB A) on the same mysql server while running a export on (DB B) . the import will hang until my export is complete ?

**Reply by davidducos** [reply]
Hi @viswaprsnl, yes, I think that it will be block

**Comment by viswaprsnl** [comment]
@davidducos do we have a flag to turn off this ? I assume that will result in inconsistent back ups ? why do we need a global lock , can we just not lock the DDL statements on just the DB thats been exported ?

**Comment by davidducos** [comment]
Hi @viswaprsnl,

We have this:
```
Lock Options
  -z, --tidb-snapshot              Snapshot to use for TiDB
  -k, --no-locks                   Do not execute the temporary shared read lock.  WARNING: This will cause inconsistent backups
  --use-savepoints                 Use savepoints to reduce metadata locking issues, needs SUPER privilege
  --no-backup-locks                Do not use Percona backup locks
  --lock-all-tables                Use LOCK TABLE for all, instead of FTWRL
  --less-locking                   Minimize locking time on InnoDB tables.
  --trx-consistency-only           Transactional consistency only
```
Depending the database vendor, it will execute different DDL locks. With --no-locks disables it.

**Reply by viswaprsnl** [reply]
@davidducos is this a possibility ? why do we need a global lock , can we just not lock the DDL statements on just the DB thats been exported ?

**Reply by davidducos** [reply]
Hi @viswaprsnl we can add an option to do not lock the DDL statements, something like --skip-ddl-locks, but take into account that the backup is going to be cancel if a ddl occurs. I thought of adding  --skip-ddl-locks in the past, but I never created the feature request, as I always thought that consistency is preferable. However, you established a good point, which is partial backups. 
Feel free to create the FR, if I don't found it in a couple of weeks, for sure I will create it.

**Reply by viswaprsnl** [reply]
Done thanks for looking into it. https://github.com/mydumper/mydumper/issues/1297 . I have couple of questions on this . How is  this is different from --no locks ?
can we not just lock the DDL statements for the particular DB and not for all the DB  sitting on the same instance ?
--skip ddl locks is only for the DB thats been exported or for all the DB in the same instance ?

**Comment by viswaprsnl** [comment]
@davidducos quick question again on the same locks  , below is the locking pattern for an export . is there a reason for us doing both DDL lock and FTWRL , can we not just use one ? older version only supports FTWRL which was fine , but doing a global ddl lock is harming other activities on the DB while export is happening ? can we put this in a flag where we can choose to use either DDL lock or FTWRL
2023-09-14 12:06:57 [INFO] - Connected to a Percona server
2023-09-14 12:06:57 [INFO] - Acquiring DDL lock
2023-09-14 12:06:57 [INFO] - Acquiring Global lock
2023-09-14 12:06:57 [INFO] - Sending Flush Table
2023-09-14 12:06:57 [INFO] - Acquiring FTWRL

**Reply by davidducos** [reply]
We need FTWRL to sync the threads, which is released after all non-innodb tables has been export.
We kept the DDL lock till the whole backup to avoid an inconsistent backup. If you are using Percona Server and GTID, feel free to try with --no-locks, at the end, it will say if it is a consistent backup or not.
